k2/csrc/fsa_utils_inl.h

```diff
@@ -0,0 +1,238 @@
+/**
+ * @brief
+ * fsa_utils_inl
+ *
+ * @note
+ * Don't include this file directly; it is included by fsa_utils.h.
+ * It contains implementation code.
+ *
+ * @copyright
+ * Copyright (c)  2020  Xiaomi Corporation (authors: Haowen Qiu)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#ifndef K2_CSRC_FSA_UTILS_INL_H_
+#define K2_CSRC_FSA_UTILS_INL_H_
+
+#ifndef IS_IN_K2_CSRC_FSA_UTILS_H_
+#error "this file is supposed to be included only by fsa_utils.h"
+#endif
+
+#include <algorithm>
+#include <limits>
+#include <utility>
+#include <vector>
+
+#include "k2/csrc/array.h"
+#include "k2/csrc/fsa.h"
+#include "k2/csrc/ragged.h"
+
+namespace k2 {
+template <typename FloatType>
+Array1<FloatType> GetForwardScores(FsaVec &fsas, Ragged<int32_t> &state_batches,
```

Comment:
BTW, should we returen Array1<double> for those forward/backward scores? no matter `FloatType` is `float` or `double`

Reply:
Dunno what you are asking, but yes, return array1.  The function will need to be called with explicit template argument.

Reply:
If implementing the template is a hassle, do just double for now.

Reply:
Sorry, I mean return `Array1<double>`. Anyway, will leave as it is as we talked offline.